src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,291 @@
+//! Polymorphization Analysis
+//! =========================
+//!
+//! This module implements an analysis of functions, methods and closures to determine which
+//! generic parameters are unused (and eventually, in what ways generic parameters are used - only
+//! for their size, offset of a field, etc.).
+
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::{
+    visit::{TyContext, Visitor},
+    Local, LocalDecl, Location,
+};
+use rustc_middle::ty::{
+    self,
+    fold::{TypeFoldable, TypeVisitor},
+    query::Providers,
+    Const, Ty, TyCtxt,
+};
+
+/// Provide implementations of queries relating to polymorphization analysis.
+pub fn provide(providers: &mut Providers<'_>) {
+    providers.unused_generic_params = unused_generic_params;
+}
+
+/// Determine which generic parameters are used by the function/method/closure represented by
+/// `def_id`. Returns a `u64` where a bit is set if the parameter with that index is unused (ie.
+/// a value of zero indicates that all parameters are used).
+fn unused_generic_params(tcx: TyCtxt<'_>, def_id: DefId) -> u64 {
+    debug!("unused_generic_params({:?})", def_id);
+
+    let generics = tcx.generics_of(def_id);
+    debug!("unused_generic_params: generics={:?}", generics);
+
+    // Exit early when there are no parameters to be unused.
+    if generics.count() == 0 {
+        return 0;
+    }
+
+    // Exit early when there is no MIR available.
+    if !tcx.is_mir_available(def_id) {
+        debug!("unused_generic_params: (no mir available) def_id={:?}", def_id);
+        return 0;
+    }
+
+    // Use a `u64` as a bitset. Starting with all ones, shift left by the number of parameters,
+    // leaving N zeros for each parameter. When a parameter is marked as used, the bit (from the
+    // left) corresponding to the parameter index will be flipped. This is the opposite of what
+    // will be returned.
+    let mut used_parameters = u64::max_value() << generics.count();
```

Comment:
(Potentially irrelevant comment, This PR looks very exciting!)  This expression will panic if `generics.count()` is larger than 63.

Reply:
Fixed!